apm-agent-core/src/test/java/co/elastic/apm/agent/report/PartialTransactionTest.java

```diff
@@ -0,0 +1,210 @@
+                assertThatJson(metadata).inPath("$.metadata.system").isObject();
+
+                String transaction = body[1];
+                assertThatJson(transaction).inPath("$.transaction.timestamp").isEqualTo(tx1.getTimestamp());
+                assertThatJson(transaction).inPath("$.transaction.id").isEqualTo(tx1.getTraceContext().getId().toString());
+                assertThatJson(transaction).inPath("$.transaction.trace_id").isEqualTo(tx1.getTraceContext().getTraceId().toString());
+                assertThatJson(transaction).inPath("$.transaction.faas.execution").isEqualTo("foo-bar-id");
+            });
+
+        tx1.deactivate();
+
+        //ensure reporting only happens after first activation
+        tx1.activate();
+        tx1.deactivate().end();
+        apmServer.verify(1, postRequestedFor(urlEqualTo("/register/transaction")));
+
+        Transaction tx2 = tracer
+            .startRootTransaction(null)
+            .withName("second-faas-transaction");
+        tx2.getFaas().withExecution("baz-id");
+        tx2.activate();
+
+        assertThat(apmServer.findAll(postRequestedFor(urlEqualTo("/register/transaction"))))
+            .hasSize(2)
+            .element(1)
+            .satisfies(req -> {
+                assertThat(req.getHeader("Content-Type")).isEqualTo("application/vnd.elastic.apm.transaction+ndjson");
+                assertThat(req.getHeader("x-elastic-aws-request-id")).isEqualTo("baz-id");
+
+                String[] body = req.getBodyAsString().split("\n");
+                assertThat(body).hasSize(2);
+
+                String metadata = body[0];
+                assertThatJson(metadata).inPath("$.metadata.service").isObject();
+                assertThatJson(metadata).inPath("$.metadata.process").isObject();
+                assertThatJson(metadata).inPath("$.metadata.system").isObject();
+
+                String transaction = body[1];
+                assertThatJson(transaction).inPath("$.transaction.timestamp").isEqualTo(tx2.getTimestamp());
+                assertThatJson(transaction).inPath("$.transaction.id").isEqualTo(tx2.getTraceContext().getId().toString());
+                assertThatJson(transaction).inPath("$.transaction.trace_id").isEqualTo(tx2.getTraceContext().getTraceId().toString());
+                assertThatJson(transaction).inPath("$.transaction.faas.execution").isEqualTo("baz-id");
+            });
+
+        tx2.deactivate().end();
+    }
+
+
+    @Test
+    public void testNoMoreReportingAfter4xx() {
+        apmServer.stubFor(post(urlEqualTo("/register/transaction")).willReturn(aResponse().withStatus(404)));
+        ServerlessConfiguration serverlessConfig = spyConfig.getConfig(ServerlessConfiguration.class);
+        doReturn(true).when(serverlessConfig).runsOnAwsLambda();
+
+        Transaction tx1 = tracer
+            .startRootTransaction(null)
+            .withName("faas-transaction");
+        tx1.getFaas().withExecution("foo-bar-id");
```

Comment:
[question] if I understand correctly the `withExecution` call makes the transaction as a FAAS transaction, in opposition to a "regular" transaction where this is not called.

Reply:
Exactly.